### PR TITLE
BUG: `optimize.curve_fit`: fix sigma dimension issue with `nan_policy="omit"`

### LINKED ISSUE #21772: BUG: optimize.curve_fit with nan_policy = "omit" fails to handle sigma properly
### Describe your issue.

Hello,
I have encountered an issue with curve_fit. It is bypassable, but I'd say it is unexpected behaviour. Generally speaking, it occurs when both of the following is true:
-- I have xdata or ydata with NaNs and use nan_policy = "omit" to handle them;
-- I have specified sigma.
In this case, I get ValueError "Sigma has incorrect shape" in optimize._minpack_py.py despite sigma being of the same size as xdata and ydata.

I checked what is going in _minpack_py.py and figured out the following.
It occurs because "nan_policy = omit" leads to the removal of NaN indices from xdata and ydata (lines 960-967), but not from sigma. Then, after line 969, the shape of sigma is checked by comparing it with ydata. And even if sigma and ydata were originally of the same size, they end up to be different after ydata truncation, throwing an error.

So, to bypass this problem, one has either to remove indices corresponding NaNs in xdata and ydata from sigma, or not using nan_policy = "omit" at all and remove NaN indices manually. I suggest improving nan_policy = "omit" in a way that it would remove the corresponding indices from sigma, if needed.

Sorry if my bug report is incorrect, I am first time doing this.

### Reproducing Code Example

```python
import numpy as np
from scipy.optimize import curve_fit

def exponential(x, a, b):
	return b * np.exp(a * x)

np.random.seed(0)
N = 90
x = np.linspace(1, 10, N, endpoint = False)
sigma = x * 0.05
y = exponential(x, 0.2, 0.5) + np.random.normal(0, sigma, N)

p0 = [0.1, 1.0]

# without nans, works perfect
popt, pcov = curve_fit(exponential, x, y, p0 = p0, sigma = sigma)
perr = np.diag(np.sqrt(pcov))
print("without nans:")
print(f"popt: {popt}")
print(f"perr: {perr}")

# with nans, have to set nan_policy = "omit"...
x[11] = np.nan
y[22] = np.nan
popt, pcov = curve_fit(exponential, x, y, p0 = p0, nan_policy = "omit")
perr = np.diag(np.sqrt(pcov))
print("with nans:")
print(f"popt: {popt}")
print(f"perr: {perr}")

#...but when using sigma, it says "sigma has incorrect shape"
x[11] = np.nan
y[22] = np.nan
popt, pcov = curve_fit(exponential, x, y, p0 = p0, sigma = sigma, nan_policy = "omit")
```

### Error message

```shell
Traceback (most recent call last):
  File "/root/Science/Spiral_shape_analysis/fit_results/../spiralcutter_advanced/scipy_test.py", line 34, in <module>
    popt, pcov = curve_fit(exponential, x, y, p0 = p0, sigma = sigma, nan_policy = "omit")
  File "/usr/local/lib/python3.10/dist-packages/scipy/optimize/_minpack_py.py", line 985, in curve_fit
    raise ValueError("`sigma` has incorrect shape.")
ValueError: `sigma` has incorrect shape.
```

### SciPy/NumPy/Python version and system information

```shell
>>> print(scipy.__version__, numpy.__version__, sys.version_info)
1.14.1 2.1.2 sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)
>>> scipy.show_config()
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /opt/_internal/cpython-3.10.14/lib/python3.10/site-packages/scipy_openblas32/include
    lib directory: /opt/_internal/cpython-3.10.14/lib/python3.10/site-packages/scipy_openblas32/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.27.dev DYNAMIC_ARCH NO_AFFINITY Zen MAX_THREADS=64
    pc file directory: /project
    version: 0.3.27.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /opt/_internal/cpython-3.10.14/lib/python3.10/site-packages/scipy_openblas32/include
    lib directory: /opt/_internal/cpython-3.10.14/lib/python3.10/site-packages/scipy_openblas32/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.27.dev DYNAMIC_ARCH NO_AFFINITY Zen MAX_THREADS=64
    pc file directory: /project
    version: 0.3.27.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.12.0
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.11
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  pythran:
    include directory: ../../tmp/pip-build-env-h_xz8lfs/overlay/lib/python3.10/site-packages/pythran
    version: 0.16.1
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /opt/python/cp310-cp310/bin/python
  version: '3.10'
```

### PR DESCRIPTION
**Pull request by SoutrikBandyopadhyay**
#### Reference issue
Closes #21772.

#### What does this implement/fix?

Fixes the bug in optimize.curve_fit where using nan_policy="omit" did not remove corresponding entries of sigma.
In this PR, I fix this bug by checking for the dimension of sigma and removing the entries where either xdata or ydata is NaN.

### CODE REVIEW

**Comment by j-bowhay** [comment]
Please could you add a regression test that fails on main but passes for this pr

**Comment by SoutrikBandyopadhyay** [comment]
@j-bowhay Added a regression test `test_curvefit_omitnan` that fails on main but passes for this PR.

File: scipy/optimize/_minpack_py.py [hunk 1]
```diff
@@ -961,9 +961,20 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
             has_nan = has_nan.any(axis=tuple(range(has_nan.ndim-1)))
             has_nan |= np.isnan(ydata)
 
+            # Store the size of ydata before purging the nan values
+            prev_ydata_size = ydata.size
+
             xdata = xdata[..., ~has_nan]
             ydata = ydata[~has_nan]
 
+            # Also omit the corresponding entries form sigma
```
[the remark this answers is not in the document]
Comment (SoutrikBandyopadhyay) [reply scipy/optimize/_minpack_py.py:970]: Oops!

File: scipy/optimize/tests/test_minpack.py [hunk 2]
```diff
@@ -804,6 +804,27 @@ def test_maxfev_and_bounds(self):
         assert_allclose(popt1, 2, atol=1e-14)
         assert_allclose(popt2, 2, atol=1e-14)
 
+    def test_curvefit_omitnan(self):
+        def exponential(x, a, b):
+            return b * np.exp(a * x)
+
+        np.random.seed(0)
```
Comment (j-bowhay) [inline scipy/optimize/tests/test_minpack.py:811]: We should avoid setting the global random state and instead create and use a generator 
```suggestion
        rng = np.random.default_rng(578285731148908)
```

Comment (SoutrikBandyopadhyay) [reply scipy/optimize/tests/test_minpack.py:811]: Thank you for pointing this out. I didn't know this.

File: scipy/optimize/_minpack_py.py [hunk 3]
```diff
@@ -961,15 +961,26 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
...
+            # Store the size of ydata before purging the nan values
+            prev_ydata_size = ydata.size
+
             xdata = xdata[..., ~has_nan]
             ydata = ydata[~has_nan]
 
+            # Also omit the corresponding entries from sigma
+            if sigma is not None:
+                sigma = np.asarray(sigma)
+                if sigma.shape == (prev_ydata_size,):
```
Comment (mdhaber) [inline scipy/optimize/_minpack_py.py:973]:
```suggestion
                if sigma.ndin == 1:
```

File: scipy/optimize/_minpack_py.py [hunk 4]
```diff
@@ -961,15 +961,26 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
...
+                    sigma = sigma[~has_nan]
+                if sigma.shape == (prev_ydata_size, prev_ydata_size):
```
Comment (mdhaber) [inline scipy/optimize/_minpack_py.py:975]:
```suggestion
                Elif sigma.ndim == 2:
```

File: scipy/optimize/_minpack_py.py [hunk 5]
```diff
@@ -961,15 +961,26 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
             has_nan = has_nan.any(axis=tuple(range(has_nan.ndim-1)))
             has_nan |= np.isnan(ydata)
 
+            # Store the size of ydata before purging the nan values
```
Comment (mdhaber) [inline scipy/optimize/_minpack_py.py:964]:
```suggestion
```

File: scipy/optimize/_minpack_py.py [hunk 6]
```diff
@@ -961,15 +961,26 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
...
+            prev_ydata_size = ydata.size
```
Comment (mdhaber) [inline scipy/optimize/_minpack_py.py:965]:
```suggestion
```

File: scipy/optimize/_minpack_py.py [hunk 7]
```diff
@@ -961,15 +961,26 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
...
+
```
Comment (mdhaber) [inline scipy/optimize/_minpack_py.py:966]:
```suggestion
```

**Review by mdhaber — COMMENTED** [review]
Do we need `prev_ydata_size`? I thought it was either a scalar, 1d array, or 2d array, in which case this can be simplified.

File: scipy/optimize/_minpack_py.py [hunk 8]
```diff
@@ -961,15 +961,26 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
...
+            # Also omit the corresponding entries from sigma
+            if sigma is not None:
+                sigma = np.asarray(sigma)
+                if sigma.shape == (prev_ydata_size,):
+                    sigma = sigma[~has_nan]
+                if sigma.shape == (prev_ydata_size, prev_ydata_size):
+                    sigma = sigma[~has_nan, ~has_nan]
+
     # Determine type of sigma
     if sigma is not None:
```
Comment (mdhaber) [inline scipy/optimize/_minpack_py.py:979]: This is no longer needed, right?

File: scipy/optimize/_minpack_py.py [hunk 9]
```diff
@@ -964,12 +964,20 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
...
+                elif sigma.ndim == 2:
+                    sigma = sigma[~has_nan, ~has_nan]
+
     # Determine type of sigma
     if sigma is not None:
         sigma = np.asarray(sigma)
 
         # if 1-D or a scalar, sigma are errors, define transform = 1/sigma
-        if sigma.size == 1 or sigma.shape == (ydata.size, ):
+        if sigma.size == 1 or sigma.shape == (ydata.size,):
```
Comment (mdhaber) [inline scipy/optimize/_minpack_py.py:980]: I see now that you were trying to match this style. I'm not sure it was done that way. To begin with. ISTM these could look at the number of dimensions, not the size. That is what the comment says, after all.

Comment (SoutrikBandyopadhyay) [reply scipy/optimize/_minpack_py.py:980]: Yeah, I understand. I think checking the `ndim` in the present case works. If, after the omission of NaNs, the sigma still differs in size from the `ydata`, the existing code will catch it and raise the appropriate error.

Comment (SoutrikBandyopadhyay) [reply scipy/optimize/_minpack_py.py:980]: On second thought, what happens when sigma differs in size with ydata even before the omission of NaNs? For example consider the case when `sigma.size = 80` and `ydata.size = 100`. In this case the `sigma = sigma[~has_nan]` will raise an IndexError. Thus I believe we need to check the shape of sigma and not just the ndim.

**Comment by mdhaber** [comment]
Isn't the only *valid* shape of `sigma` 1, `n,`, or `(n, n)`? If so, differences in shape should be caught in some sort of input validation that raises an appropriate error. If it's not there already, I would not add it in this PR, though. I would also not recommend sprinkling it in throughout the logic of the code.

**Comment by SoutrikBandyopadhyay** [comment]
In that case, the ndim check suffices.

**Comment by mdhaber** [comment]
Yeah, I can clean it up later. An Index error for invalid input is much better than the current error for valid input.

File: scipy/optimize/tests/test_minpack.py [hunk 10]
```diff
@@ -804,6 +804,28 @@ def test_maxfev_and_bounds(self):
         assert_allclose(popt1, 2, atol=1e-14)
         assert_allclose(popt2, 2, atol=1e-14)
 
+    def test_curvefit_omitnan(self):
```
Comment (mdhaber) [inline scipy/optimize/tests/test_minpack.py:807]: It looks like this tests whether the answer with NaNs omitted is approximately equal to the answer with all the original data for the case of 1D `sigma`.

Ideally, we'd test whether the answer with NaNs omitted is *correct* for *all* cases of `sigma`: 0D, 1D, and 2D.

We can test whether the answer is correct by omitting the NaNs manually, maybe using an approach other than a logical mask.

I can't test this code right now, but I'm thinking something like:
```python3
i_x = rng.integers(N, size=5)
i_y = rng.integers(N, size=5)

x[i_x] = np.nan
y[i_y] = np.nan
res = curve_fit(exponential, x, y, p0=p0, sigma=sigma, nan_policy="omit")

i_delete =np.unique(np.concat((i_x, i_y)))
x = np.delete(x, i_delete)
y = np.delete(y, i_delete) 
if sigma.ndim == 1:
    sigma = np.delete(sigma, i_delete)
elif sigma.ndim == 2:
    sigma = np.delete(sigma, i_delete, axis=0)
    sigma = np.delete(sigma, i_delete, axis=1)
ref = curve_fit(exponential, x, y, p0=p0, sigma=sigma)

assert_equal(res, ref)
```

We can parametrize the test (`pytest.mark.parametrize('sigma', ...)`) over the different possibilities of `sigma`.

Does that look like a stronger test to you, @j-bowhay @lucascolley?

**Review by mdhaber — APPROVED** [review]
This is close; I'd just like for the test to be stronger. I will follow-up with a PR to improve the input validation. (Or would you like to, @SoutrikBandyopadhyay? The idea would be to validate the shape of `sigma` earlier, before dealing with `nan_policy` so that we get a nice error message instead of the `IndexError` you mentioned.)

File: scipy/optimize/tests/test_minpack.py (hunk 10 above)
Comment (SoutrikBandyopadhyay) [reply scipy/optimize/tests/test_minpack.py:807]: Yes, this looks like a good way to handle different sigma dimensions. I am on it.

**Comment by SoutrikBandyopadhyay** [comment]
I am happy to work on this. Could you please create an issue for the same? Additionally, what kind of error message would we like to throw?

File: scipy/optimize/_minpack_py.py [hunk 11]
```diff
@@ -970,7 +970,8 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
                 if sigma.ndim == 1:
                     sigma = sigma[~has_nan]
                 elif sigma.ndim == 2:
-                    sigma = sigma[~has_nan, ~has_nan]
+                    sigma = sigma[~has_nan, :]
```
Comment (SoutrikBandyopadhyay) [inline scipy/optimize/_minpack_py.py:973]: I caught this subtle error while creating a test for the 2D case.
The line `sigma = sigma[~has_nan,~has_nan]` was wrong, and it produced a 1D array. This commit fixes this bug and the sigma is now the correct shape and contains the correct values.

Comment (SoutrikBandyopadhyay) [reply scipy/optimize/_minpack_py.py:973]: Now the code handles all the cases of sigma perfectly.

**Comment by mdhaber** [comment]
Thanks @SoutrikBandyopadhyay!

>  Could you please create an issue for the same? Additionally, what kind of error message would we like to throw?

You're welcome to open an issue about it; it doesn't need to come from me. I'm also comfortable if you'd just like to open a PR without an issue.

The error message would be similar to what it was before this PR when the shape of `sigma` is invalid. It would just get raised earlier so you don't get an index error.

**Comment by SoutrikBandyopadhyay** [comment]
Thank you @mdhaber. I am on it!

**Pull request merged**